roachtest: export openmetrics from tpccbench

sambhav-jain-16 opened the pull request:
This change adds capability to emit openmetrics in `tpccbench`. The tests calculate tpmc in its result calculation using hdrhistogram json. Therefore, tpccbench will emit hdr json from workload binary and then themselves convert them into openmetrics in the test itself.

This change also aims to refactor `exporter.Close()` from
`CreateStatsFileFromExporter` to a each roachtest and use defer to make
the code more clean and easy to understand.


Epic: none

Release note: None

sambhav-jain-16 committed:
roachtest: export openmetrics from tpccbench

This change adds capability to emit openmetrics in `tpccbench`. The
tests calculate tpmc in its result calculation using hdrhistogram json.
Therefore, tpccbench will emit hdr json from workload binary and then
themselves convert them into openmetrics in the test itself.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/135734)
<!-- Reviewable:end -->

sambhav-jain-16 commented:
Tested this here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Roachtest_Openmetrics_Gce/17843065?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true

and verified the metrics are coming in grafana

nameisbhaskar commented on pkg/cmd/roachtest/tests/tpcc.go:
```diff
@@ -1866,3 +1886,27 @@ func getTpccLabels(
+		}
+	}
+
+	if err := exporter.Close(nil); err != nil {
```
Please move this to defer right after Init.

sambhav-jain-16 replied:
Actually, there's already an `exporter.Close()` in `CreateStatsFileInClusterFromExporterWithPrefix`, so I think I'll remove this already

sambhav-jain-16 committed:
roachtest: refactor and code cleanup in openmetrics emission in roachtests

1. This change aims to refactor `exporter.Close()` from
`CreateStatsFileFromExporter` to a each roachtest and use defer to make
the code more clean and easy to understand.
2. There were some code smell and nit errors in `statsexporter` package.
This change also fixes those.

Epic: none

Release note: None

(cherry picked from commit bd62d73e6a8ae730ecba09412aa473a8a2a6a3f6)

sambhav-jain-16 commented:
Verified changes here !

srosenberg commented on pkg/cmd/roachtest/roachtestutil/utils.go:
```diff
@@ -94,11 +94,18 @@ func GetBenchmarkMetricsFileName(t test.Test) string {
+	return CreateWorkloadHistogramExporterWithLabels(t, c, nil)
+}
+
+// CreateWorkloadHistogramExporterWithLabels CreateWorkloadHistogramExporter creates a exporter.Exporter based on the roachtest parameters with additional labels
```
Nit: remove `CreateWorkloadHistogramExporter`

srosenberg commented on pkg/cmd/roachtest/tests/tpcc.go:
```diff
@@ -1716,9 +1720,36 @@ func runTPCCBench(ctx context.Context, t test.Test, c cluster.Cluster, b tpccBen
+
+				// This roachtest uses the stats.json emitted from hdr histogram to compute Tpmc and show it in the run log
+				// Since directly emitting openmetrics and computing Tpmc from it is not supported, it is better to convert the
+				// stats.json emitted and convert it into openmetrics in the test itself and upload it to the cluster
```
Nit: second  occurrence of `convert` is redundant

srosenberg commented on pkg/cmd/roachtest/tests/tpcc.go:
```diff
@@ -1866,3 +1897,19 @@ func getTpccLabels(
+
+// This function converts exporter.SnapshotTick to openmetrics into a buffer
+func getHistogramOpenmetrics(
+	exporter exporter.Exporter, snapshots map[string][]exporter.SnapshotTick, perfBuf *bytes.Buffer,
```
`perfBuf` is unused

sambhav-jain-16 replied:
Oops, missed this!
Thanks for pointing out !

srosenberg commented on pkg/cmd/roachtest/tests/tpcc.go:
```diff
@@ -1716,9 +1720,36 @@ func runTPCCBench(ctx context.Context, t test.Test, c cluster.Cluster, b tpccBen
+						}
+					}()
+
+					if err := getHistogramOpenmetrics(exporter, snapshots, perfBuf); err != nil {
```
Rename to `exportOpenMetrics` since the helper isn't just "getting" data? It's essentially performing a conversion and outputting it to `perfBuf`.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/utils.go:
```diff
@@ -118,10 +125,27 @@ func CreateStatsFileInClusterFromExporter(
+		return "", errors.New("perf buffer is nil")
+	}
+	// Close and flush any existing data in the buffer
 	if err := exporter.Close(nil); err != nil {
```
Seems weird to do it in the callee as opposed to the caller. The caller should take care of closing the exporter, thus ensuring that the output buffer is "whole". This is just plain old separation of concerns. Thus, the callee need not be aware of the `exporter`.

sambhav-jain-16 replied:
This is something that I'm fixing in this PR https://github.com/cockroachdb/cockroach/pull/135821.

If I remove it in this PR, the other roachtests will fail. One way is that I merge the above mentioned PR into this PR, but I wanted this PR to be only for tpccbench

srosenberg replied:
I think if you rebase this PR on top of https://github.com/cockroachdb/cockroach/pull/135821, GH will hide base changes, or at least Reviewable will. It would make it easier to review the _union_ of the changes.

sambhav-jain-16 replied:
Makes sense.
Let me include the changes in this changes itself.

sambhav-jain-16 replied:
I have merged the PRs into this one.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/utils.go:
```diff
@@ -110,29 +117,50 @@ func CreateWorkloadHistogramExporter(t test.Test, c cluster.Cluster) exporter.Ex
+		}
+		return nil
+	}); err != nil {
+		t.Errorf("failed to close exporter: %v", err)
```
Maybe "failed to export" since "close" in this context means executing a function which uploads a file.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/utils.go:
```diff
@@ -110,29 +117,50 @@ func CreateWorkloadHistogramExporter(t test.Test, c cluster.Cluster) exporter.Ex
 
-func CreateStatsFileInClusterFromExporter(
+// CreateStatsFileInNode creates stats file from buffer in the node
+func CreateStatsFileInNode(
```
`UploadPerfStats`?

srosenberg commented on pkg/cmd/roachtest/tests/decommissionbench.go:
```diff
@@ -675,6 +675,15 @@ func runDecommissionBench(
 
+	defer func() {
+		if err := exporter.Close(func() error {
+			uploadPerfArtifacts(ctx, t, c, benchSpec, pinnedNode, workloadNode, perfBuf)
```
Are we uploading it twice? L506 uploads it to `pinnedNode`.

sambhav-jain-16 replied:
It uploads only once, since we go through `uploadPerfArtifacts`. `exporter.Close` doesn't upload the file.

srosenberg replied:
Thanks, I was confused by the truncated diff. I think GH UI could benefit from placing a function name above the truncated code block.

sambhav-jain-16 commented:
TFTR !

bors r=@srosenberg

sambhav-jain-16 commented:
Successful run here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Roachtest_Openmetrics_Gce/17860410?buildTab=artifacts#%2Ftpcc%2Fheadroom%2Fn4cpu16%2Frun_1%2Fartifacts.zip
